actors/builtin/market/market_test.go

```diff
@@ -1 +1,252 @@
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, client))
+		}
+	})
+
+}
+
+func TestWithdrawBalance(t *testing.T) {
+	actor := marketActorTestHarness{t: t}
+	marketActor := tutil.NewIDAddr(t, 100)
+	provider := tutil.NewIDAddr(t, 101)
+	worker := tutil.NewIDAddr(t, 102)
+	client := tutil.NewIDAddr(t, 103)
+
+	var st market.State
+
+	builder := mock.NewBuilder(context.Background(), marketActor).
+		WithBalance(abi.NewTokenAmount(100), abi.NewTokenAmount(0)).
+		WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+	t.Run("fails with a negative withdraw amount", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		params := market.WithdrawBalanceParams{
+			ProviderOrClientAddress: provider,
+			Amount:                  abi.NewTokenAmount(-1),
+		}
+
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			rt.Call(actor.WithdrawBalance, &params)
+		})
+
+		rt.Verify()
+	})
+
+	t.Run("withdraws from provider escrow funds", func(t *testing.T) {
+		rt := builder.WithBalance(abi.NewTokenAmount(100), abi.NewTokenAmount(0)).Build(t)
+		actor.constructAndVerify(rt)
+		actor.addProviderFunds(rt, provider, worker, abi.NewTokenAmount(20))
+
+		rt.GetState(&st)
+		assert.Equal(t, abi.NewTokenAmount(20), st.GetEscrowBalance(rt, provider))
+
+		rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+		actor.expectProviderControlAddress(rt, provider, worker)
+		rt.ExpectSend(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, abi.NewTokenAmount(0), nil, exitcode.Ok)
+
+		withdrawAmount := abi.NewTokenAmount(1)
+		rt.ExpectSend(provider, builtin.MethodSend, nil, withdrawAmount, nil, exitcode.Ok)
+
+		params := market.WithdrawBalanceParams{
+			ProviderOrClientAddress: provider,
+			Amount:                  withdrawAmount,
+		}
+
+		rt.Call(actor.WithdrawBalance, &params)
+
+		rt.Verify()
+
+		rt.GetState(&st)
+		assert.Equal(t, abi.NewTokenAmount(19), st.GetEscrowBalance(rt, provider))
+	})
+
+	t.Run("withdraws from non-provider escrow funds", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+		actor.addParticipantFunds(rt, client, abi.NewTokenAmount(20))
+
+		rt.GetState(&st)
+		assert.Equal(t, abi.NewTokenAmount(20), st.GetEscrowBalance(rt, client))
+
+		withdrawAmount := abi.NewTokenAmount(1)
+
+		// setup WithdrawBalance call
+		rt.SetCaller(client, builtin.AccountActorCodeID)
+		rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+		params := market.WithdrawBalanceParams{
+			ProviderOrClientAddress: client,
+			Amount:                  withdrawAmount,
+		}
+
+		rt.ExpectSend(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, abi.NewTokenAmount(0), nil, exitcode.Ok)
+		rt.ExpectSend(client, builtin.MethodSend, nil, withdrawAmount, nil, exitcode.Ok)
+
+		rt.Call(actor.WithdrawBalance, &params)
+
+		rt.Verify()
+
+		rt.GetState(&st)
+		assert.Equal(t, abi.NewTokenAmount(19), st.GetEscrowBalance(rt, client))
+	})
```

Comment:
Please add cases for:
- miner worker account used as deal client
- fail due to insufficient balance

Add a placeholder/todo for
- provider/client fail with sufficient balance, but locked (requires deals)